javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/InstrumentationModuleClassLoader.java

```diff
@@ -0,0 +1,256 @@
+    }
+    // for classes use the same precedence as in loadClass
+    ClassCopySource injected = getInjectedClass(className);
+    if (injected != null) {
+      return injected.getUrl();
+    }
+    URL fromAgentCl = agentOrExtensionCl.getResource(resourceName);
+    if (fromAgentCl != null) {
+      return fromAgentCl;
+    }
+    return instrumentedCl.getResource(resourceName);
+  }
+
+  @Override
+  public Enumeration<URL> getResources(String resourceName) throws IOException {
+    String className = resourceToClassName(resourceName);
+    if (className == null) {
+      return super.getResources(resourceName);
+    }
+    URL resource = getResource(resourceName);
+    List<URL> result =
+        resource != null ? Collections.singletonList(resource) : Collections.emptyList();
+    return Collections.enumeration(result);
+  }
+
+  @Nullable
+  private static String resourceToClassName(String resourceName) {
+    if (!resourceName.endsWith(".class")) {
+      return null;
+    }
+    String className = resourceName;
+    if (className.startsWith("/")) {
+      className = className.substring(1);
+    }
+    className = className.replace('/', '.');
+    className = className.substring(0, className.length() - ".class".length());
+    return className;
+  }
+
+  @Nullable
+  private ClassCopySource getInjectedClass(String name) {
+    ClassCopySource alwaysInjected = ALWAYS_INJECTED_CLASSES.get(name);
+    if (alwaysInjected != null) {
+      return alwaysInjected;
+    }
+    return additionalInjectedClasses.get(name);
+  }
+
+  private Class<?> defineClassWithPackage(String name, byte[] bytecode) {
+    int lastDotIndex = name.lastIndexOf('.');
+    if (lastDotIndex != -1) {
+      String packageName = name.substring(0, lastDotIndex);
+      safeDefinePackage(packageName);
+    }
+    return defineClass(name, bytecode, 0, bytecode.length, PROTECTION_DOMAIN);
+  }
+
+  private void safeDefinePackage(String packageName) {
+    if (findPackage(packageName) == null) {
+      try {
+        definePackage(packageName, null, null, null, null, null, null, null);
+      } catch (IllegalArgumentException e) {
+        // Can happen if a parent classloader attempts to define the same package at the same time
+        // in Java 8
+        if (findPackage(packageName) == null) {
+          // package still doesn't exist, the IllegalArgumentException must be for a different
+          // reason than a race condition
+          throw e;
+        }
+      }
+    }
+  }
+
+  /**
+   * Invokes {@link #getPackage(String)} for Java 8 and {@link #getDefinedPackage(String)} for Java
+   * 9+.
+   *
+   * <p>Package-private for testing.
+   *
+   * @param name the name of the package find
+   * @return the found package or null if it was not found.
+   */
+  @SuppressWarnings({"deprecation", "InvalidLink"})
+  Package findPackage(String name) {
+    try {
+      return (Package) FIND_PACKAGE_METHOD.invokeExact((ClassLoader) this, name);
```

Comment:
This is failing with 
```
2023-08-21T08:32:40.2006368Z InstrumentationModuleClassLoaderTest > checkClassLookupPrecedence(Path) FAILED
2023-08-21T08:32:40.2007484Z     java.lang.IllegalStateException: java.lang.invoke.WrongMethodTypeException: expected (InstrumentationModuleClassLoader,String)Package but found (ClassLoader,String)Package
2023-08-21T08:32:40.2008886Z         at io.opentelemetry.javaagent.tooling.instrumentation.indy.InstrumentationModuleClassLoader.findPackage(InstrumentationModuleClassLoader.java:225)
2023-08-21T08:32:40.2010793Z         at io.opentelemetry.javaagent.tooling.instrumentation.indy.InstrumentationModuleClassLoader.safeDefinePackage(InstrumentationModuleClassLoader.java:196)
2023-08-21T08:32:40.2012626Z         at io.opentelemetry.javaagent.tooling.instrumentation.indy.InstrumentationModuleClassLoader.defineClassWithPackage(InstrumentationModuleClassLoader.java:190)
2023-08-21T08:32:40.2014311Z         at io.opentelemetry.javaagent.tooling.instrumentation.indy.InstrumentationModuleClassLoader.loadClass(InstrumentationModuleClassLoader.java:98)
2023-08-21T08:32:40.2016186Z         at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
2023-08-21T08:32:40.2017401Z         at io.opentelemetry.javaagent.tooling.instrumentation.indy.InstrumentationModuleClassLoaderTest.checkClassLookupPrecedence(InstrumentationModuleClassLoaderTest.java:134)
2023-08-21T08:32:40.2018230Z 
2023-08-21T08:32:40.2018355Z         Caused by:
2023-08-21T08:32:40.2019060Z         java.lang.invoke.WrongMethodTypeException: expected (InstrumentationModuleClassLoader,String)Package but found (ClassLoader,String)Package
2023-08-21T08:32:40.2019868Z             at java.lang.invoke.Invokers.newWrongMethodTypeException(Invokers.java:298)
2023-08-21T08:32:40.2020731Z             at java.lang.invoke.Invokers.checkExactType(Invokers.java:309)
2023-08-21T08:32:40.2021992Z             at io.opentelemetry.javaagent.tooling.instrumentation.indy.InstrumentationModuleClassLoader.findPackage(InstrumentationModuleClassLoader.java:223)
2023-08-21T08:32:40.2022776Z             ... 5 more
``` in the tests; perhaps `invoke` should be used instead?

Reply:
Interesting one. With the Java 9+ method `getDefinedPackage` the argument type must be `ClassLoader`, that's why I added the cast in first place.
I'll change it to `invoke` and let the JVM take care of casting.